### DISCUSSION TITLE
Replica synchronization delay is serious

### DISCUSSION THREAD

**Discussion by HwiLu** [ClickHouse/ClickHouse#58494] [Q&A]
Our clickhouse cluster is a three-replication  cluster, and the version is 22.8.15.

Since this week, I have found that the data delay between the other two replicas and the first replica data is obvious, as shown in the figure below, all our data is written from the first node.
![image](https://github.com/ClickHouse/ClickHouse/assets/25824870/0376e85e-e7fe-4f20-b232-9c09bc76df0e)

I check the ck log for more info, and some logs that I think is helpful is below 
```
2024.01.04 09:50:01.728156 [ 28803 ] {} <Debug> dws_realtime_factor.dys_kline_5_min (ReplicatedMergeTreeQueue): Pulling 1 entries to queue: log-0000059832 - log-0000059832
2024.01.04 09:50:03.560627 [ 28826 ] {} <Debug> dws_realtime_factor.dys_kline_5_min (ReplicatedMergeTreeQueue): Pulling 1 entries to queue: log-0000059833 - log-0000059833
2024.01.04 09:50:11.347184 [ 28740 ] {} <Debug> dws_realtime_factor.dys_kline_5_min (ReplicatedMergeTreeCleanupThread): Removed 2 old log entries: log-0000059822 - log-0000059823
...
2024.01.04 09:50:28.085725 [ 28705 ] {} <Debug> dws_realtime_factor.dys_kline_5_min (0080dc6d-c70f-465f-8ad2-c450f00883c7): Fetching part 20240104_15_15_0 from /clickhouse/tables/xxxx/dws_realtime_xxx/dys_kline_5_min/1/replicas/192.168.xx.21
2024.01.04 09:50:28.198404 [ 28705 ] {} <Debug> dws_realtime_factor.dys_kline_5_min (0080dc6d-c70f-465f-8ad2-c450f00883c7): Fetched part 20240104_15_15_0 from /clickhouse/tables/xxxx/dws_realtime_xxx/dys_kline_5_min/1/replicas/192.168.xx.21
2024.01.04 09:50:28.342318 [ 28705 ] {} <Debug> dws_realtime_factor.dys_kline_5_min (0080dc6d-c70f-465f-8ad2-c450f00883c7): Fetching part 20240104_16_16_0 from /clickhouse/tables/xxxx/dws_realtime_xxx/dys_kline_5_min/1/replicas/192.168.xx.21
2024.01.04 09:50:28.406036 [ 28705 ] {} <Debug> dws_realtime_factor.dys_kline_5_min (0080dc6d-c70f-465f-8ad2-c450f00883c7): Fetched part 20240104_16_16_0 from /clickhouse/tables/xxxx/dws_realtime_xxx/dys_kline_5_min/1/replicas/192.168.xx.21

```

The table is updated twice every 5 minutes, The batch of data at 09:50 seems that clikhouse has synchronized the entries logs updated twice at 09:50:03, but it started downloading the partition at 09:50:28 which  waited for 25 seconds. I think this may be the reason for the delay but I don't know if it is correct.

When I first met this problem , the query `SELECT * FROM system.replication_queue` shows a lot of `Not executing fetch of part all_12377809_12377809_0 because 8 fetches already executing, max 8.`
and then I tried increasing `background_fetches_pool_size` from the default 8 to 24, but today I see no improvement.

After the adjustment, I used the following script to monitor ck

```sql
Select * FROM (Select metric, value from system.metrics where metric ilike '%Fetches%'), (select fqdn() h, is_currently_executing, count(*) from system.replication_queue
                 where  type = 'GET_PART'
                 group by h, is_currently_executing)
```
`for i in `seq 1 100000`;do echo `date`;clickhouse-client  < zz.sql ;echo;sleep .3 ;done`

It shows

![企业微信截图_1704350202442](https://github.com/ClickHouse/ClickHouse/assets/25824870/8cfa8ef9-5163-49dd-92b2-459c66a1a244)

but no log like `Not executing fetch of part all_12377809_12377809_0 because 8 fetches already executing, max 8.` could by found in 
`system.replication_queue`.

I think the slow start of the fetch part process has nothing to do with zookeeper, because downloading data does not need to go through zookeeper. But I found that zk's waiting time is very long and there are many blocked requests

![image](https://github.com/ClickHouse/ClickHouse/assets/25824870/dab29632-829c-406a-af92-1e0b9a4c593a)

![image](https://github.com/ClickHouse/ClickHouse/assets/25824870/a3400eb9-d467-4884-9770-8bb9b2972423)

![image](https://github.com/ClickHouse/ClickHouse/assets/25824870/0b6927cc-ee39-40b3-9909-629dbbcd6048)

By the way, the load on our servers is very low.
![image](https://github.com/ClickHouse/ClickHouse/assets/25824870/00368fba-d344-408a-94aa-d29841daeb4e)

I've been searching for a long time but haven't found the reason. I also want to know how to optimize it，I hope the community could help me. Thanks.

**Comment by bprakash1105** [comment]
Hi @HwiLu,
I'm also having a similar synchronization delay, were you able to resolve the issue?
If not, maybe you can join the **[https://clickhouse.com/slack](https://clickhouse.com/slack)**.
I raised this issue in a Slack thread: https://altinitydbworkspace.slack.com/archives/C02K1MWEK2L/p1723197707388749?thread_ts=1723188903.300329&cid=C02K1MWEK2L

**Comment by karthik-thiyagarajan** [comment]
Do we have any solution for this ? I see similar problems as well in our cluster.

**Reply by bprakash1105** [reply]
Hi @karthik-thiyagarajan, what version are you using?

**Reply by karthik-thiyagarajan** [reply]
24.12.3.47
